core/src/main/java/io/jenkins/pluginhealth/scoring/probes/AbstractOpenIssuesProbe.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2023 Jenkins Infra
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.jenkins.pluginhealth.scoring.probes;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.BiFunction;
+
+import io.jenkins.pluginhealth.scoring.model.Plugin;
+import io.jenkins.pluginhealth.scoring.model.ProbeResult;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonMappingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.kohsuke.github.GHRepository;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.client.RestTemplate;
+
+public abstract class AbstractOpenIssuesProbe extends Probe {
```

Comment:
Hey @Jagrutiti,

There is something wrong with that class which comes from its design.

On purpose, I won't give you more indication for now, to let you figure it out by yourself.

Please have a look at other similar classes in the project, and please read more about the purpose of `abstract` classes, and the general purpose of abstraction in OOP. 

You already implemented similar kind of classes in previous PRs, but perhaps the extra guidance we provided made that you did not get to understand the reason why we recommended this design, which is why this time I won't share much details (yet) and let you take some time to think about the design choices and understand the reason for them.

Reply:
Hi @aneveux ,

I did read about implementing an `abstract` class in OOPS. 

The articles spoke of data abstraction and functional abstraction. In our case I believe, we are doing functional abstraction. 

Having two different methods is important here because the implementation to count issues from both `JIRA` and `GitHub`  have nothing in common.

I had done a similar implementation in `AbstractDependencyBotConfigurationProbe` but in that case `RenovateProbe` and `DependabotProbe` had the same job. Look for files only with different names.

The next point that I come across is why `abstract` classes are used. They are designed for:
-  hide complex functionality
- code reusability

The abstract classes that I implemented did provide code reusability and their functionalities were hidden from the child class. They only had to implement the abstract class.

In the `NumberofOpenIssuesProbe` code reusability is not possible because the libraries and API used in both cases are different. 

About the design issues, `JIRA_HOST` and `restTemplate` are used in only case of `JIRA` issues. 

Maybe I should change the `JIRA_HOST` to `HOST_URL` and that should be assigned by all the sub-classes. 

Maybe `JIRA_HOST` and `restTemplate`  should be assigned using a constructor in the subclasses?

Reply:
> Having two different methods is important here because the implementation to count issues from both JIRA and GitHub have nothing in common.

Let me just say, what if tomorrow there is a third issue tracker used in the Jenkins project, what would we have to add or changed in here to make it fit? 

In OOP, it should require a minimal addition, very specific. 

If you look at the DependencyBot abstract class, how are the different bot handled?

Reply:
Hey @Jagrutiti 👋 

Sorry for the late reply, I'm out of the office right now and spend less time on the computer 😝 

Adrien actually gave a nice indication regarding the design: what happens if you had to add a third issue tracker?

The idea behind code abstraction is that the `abstract` class that you write should not know at all about its possible implementations. In your case, not only your `abstract` class knows about Jira and GitHub, but it even provides code for it.

An `abstract` class should contain all the code that is supposed to be common between all of its subclasses, and no more. If there is some code that is used by only one of the subclasses, it is rarely a good fit for the `abstract` class. 

That being said though, maybe while implementing your code you will notice that there is nothing to share between the two classes that you write because they are too different from each other, and in that case, it would be better to challenge the need for abstraction, and simply write 2 classes that are doing their job. 

Introducing abstraction just for the sake of it just makes the source code more complicated, hence less readable.

It is tempting to want to use good practices like abstraction while writing code when we manipulate things that seem similar, but it is important to check that they are indeed similar.

Hope that helps you understand my concern with the class design.

Reply:
Thank you for explaining @aneveux.

In past cases when I used `abstract` classes. 99% of the code was common. 

In this case, there was no common code. Maybe I should have asked this in the meeting itself but it did not occur to me then.

I should have taken inspiration from `doApply` abstract method in `Probe` class. But I did not understand the "why" part until I made this blunder.

Reply:
That's alright @Jagrutiti =) We learn by making mistakes, and if you never experienced such a case before, it would have been difficult to think about it and ask a question during the meeting.

If you understand it better now, then that's great =)